### PR TITLE
issue-2041 fix compression

### LINKED ISSUE #2041: Сжатие данных: проблемы и недоработки
## Недоделано по ТЗ
- алгоритм не должен советовать сжимать поля, которые находятся в индексе
- алгоритм не должен предлагать сжимать уже сжатые поля
- если сжатие принесет эффективность менее, чем на 10 процентов, то мы не показываем пользователю подсказку
- алгоритм должен проверять кластер на отсутствие стораджа с забитостью в 80% или более. Если есть хоть один — не давать запускать алгоритм.

## Наблюдаемые проблемы
- при включении на двух репликасетах и датасетае в 1000 записях длинной в 100кб наблюдается "моргание кластера". Это требуется исключить

### PR DESCRIPTION
**Pull request by panticonur**
Fix compression:
- freezing instances
- compress in index fileds

I didn't forget about

- [ ] Tests
- [x] Changelog
- [ ] Documentation

Close #2041

### CODE REVIEW

File: cartridge/lua-api/compression.lua [hunk 1]
```diff
@@ -134,73 +135,96 @@ function _G.__cartridgeGetStorageCompressionInfo(_)
...
-                        index_space:drop()
+                    local exist_in_compressed_space = compressed_space:get(tmp_tuple)
+                    if exist_in_compressed_space == nil then
+                        index_space:insert(tmp_tuple)
+                        fiber.yield()
+                        table.insert(tmp_tuple, full_tuple[fieldno])
+                        uncompressed_space:insert(tmp_tuple)
+                        fiber.yield()
+                        compressed_space:insert(tmp_tuple)
+                        fiber.yield()
```
Comment (yngvar-antonsson) [inline cartridge/lua-api/compression.lua:207]: I propose to put it to a single transaction (or at least the first two inserts) and test if there is any performance difference.

File: cartridge/lua-api/compression.lua [hunk 2]
```diff
@@ -134,73 +135,96 @@ function _G.__cartridgeGetStorageCompressionInfo(_)
...
+            end
 
-                    local field_in_index = false
+            local unique_index = nil
+            for _, index in pairs(space.index) do
+                if index.unique then
+                    unique_index = index
+                    break
+                end
+            end
```
Comment (yngvar-antonsson) [inline cartridge/lua-api/compression.lua:149]: The primary index (index[0]) should be unique, so you can remove this check.

File: cartridge/lua-api/compression.lua [hunk 3]
```diff
@@ -134,73 +135,96 @@ function _G.__cartridgeGetStorageCompressionInfo(_)
...
-                            compression_percentage =
-                                (compressed_space:bsize() - index_space:bsize()) * 100 /
-                                (uncompressed_space:bsize() - index_space:bsize() + 1),
-                        }
-                        table.insert(space_compression_info, field_compression_info)
-
-                        compressed_space:drop()
-                        uncompressed_space:drop()
-                        index_space:drop()
+                    local exist_in_compressed_space = compressed_space:get(tmp_tuple)
```
Comment (yngvar-antonsson) [inline cartridge/lua-api/compression.lua:199]: Does `compressed_space:get` return an uncompressed or compressed tuple?
If there is an uncompression process after get, I think you should avoid it.

File: cartridge/lua-api/compression.lua (hunk 1 above)
Comment (panticonur) [reply cartridge/lua-api/compression.lua:207]: So I had tested case with inserts in transaction. But there is no performance.

File: cartridge/lua-api/compression.lua (hunk 3 above)
Comment (panticonur) [reply cartridge/lua-api/compression.lua:199]: In this line there is difference to use `compressed_space ` or 'uncompressed_space'. So I Will Change there to 'uncompressed_space'.

File: cartridge/lua-api/compression.lua [hunk 4]
```diff
@@ -212,6 +248,8 @@ function _G.__cartridgeGetStorageCompressionInfo(_)
     }, nil
 end
 
+--require('jit').off(_G.__cartridgeGetStorageCompressionInfo)
+
```
Comment (yngvar-antonsson) [inline cartridge/lua-api/compression.lua:252]:
```suggestion
--require('jit').off(_G.__cartridgeGetStorageCompressionInfo)

```

**Review by yngvar-antonsson — APPROVED** [review]

**Pull request merged**